GridTools-Glossary.md

```diff
@@ -26,12 +26,20 @@ A weather and climate model developed by DWD and MPI-M build on a icosahedral gr
 
 TODO disambiguate from compute domain or make it an alias.
 
+### Local Field
+
+An array located on a [grid point](#Grid-point) with one entry per neighbor location.
+
 ### Origin
 
 ### Pentagon problem
 
 In the ICON grid, the sphere is built from rhomboids. Within each rhomboid the mesh and dual-mesh are fully structured (triangles, each vertex has exactly 6 edges). At corner points where the rhomboids are patched together, each vertex has only 5 neighbors. TODO add picture.
 
+### Sparse Field
+
+In the context of [unstructured grids](#Unstructured-Grid), a sparse field, is a field defined on the neighbors of each [grid point](#grid-point). The restriction of a sparse field to a single [grid point](#grid-point) is a [local field](#Local-Field). The name originates from the sparse nature of the second / neighbor dimension.
```

Comment:
A field whose entries are references to topological entities around them is a sort of incidence matrix, which is a rectangular matrix that is indeed sparse. In this case though, the entries are implicitly assigned to the neighbor entities according to an ordering of those entities. This ordering is arbitrary, but they are stored compactly, at least in principles. There may be "holes" because there are differences in the number of neighbor entities. [_We often say, we may be limiting these fields to entities with fixed number of neighbors (edges to vertices, cells to edges), do we know if this is possible? If not, is the count of how many neighbors there are in another field?_]

Could you have a field with local fields that is not sparse? Do we need two definitions?

Reply:
>Could you have a field with local fields that is not sparse? Do we need two definitions?

In theory, yes. By defining the neighborhood to be everything. It's probably not very useful...
A more useful view (in my opinion) is to have a dense field which is not constructed from a LocalField, but has on each gridpoint an entry for each gridpoint. Since this is only a different view, it's useless for the same reason...

Reply:
but this has O(N^2) columns, when N is the size of the mesh... Would it make sense since this could be huge (presumably in a domain decomposed application this could be something like O(N^2/P) in each node

Reply:
Indeed the sparsity pattern of a sparse field, viewed as a matrix, is equal to the corresponding incidence matrix. As you said, the entries are implicitly assigned, but it is important to note that this assignment is well defined by the order, which is not arbitrary, not even on the concept level. This is the key result of the _mesh/no mesh_ discussions. For incidence relations (relations between "locations" and it's constituents) this was clear from the beginning. To give an example: The vertices of a cell are specified in counter-clockwise order in order to easily compute geometric properties, like the volume. What was new, is that this ordering also applies to adjacency relations. The diamond pattern occuring in ICON is an example of this, where the ordering of the neighbors is used on the concept level. Our conclusion from this was to give the user direct control over the ordering by simply respecting the ordering as supplied in the connectivities (in the future this could be emphasized here by defining connectivities, sparse fields etc. formally and using tuples instead of sets, hence reflecting the ordering at the concept level). To sum things up: the ordering from our perspective is not arbitrary, but something the domain scientist may rely on.

At this point we do not know whether sparse fields can be limited to locations with a fixed number of incident/adjacent cells. What we know is that we should at least be able to distinguish between the two cases (@anstaf new USID IR proposal introduced the `has_max_neighbors` property for that reason), but we don't and should not rely on it in the higher level IRs. In the current process of understanding the roles of the different IRs the NIR is actually classified by exactly this:

> Neighbor IR is not aware of the number of neighbors of a location. We should be able to write (theoretically) a
> backend where the number of neighbors is dynamic.

To come back to your question. Yes, the amount of neighbors is known and it is known for every local field implicitly by its reference to the connectivity (if you know what neighbors you have you also know how many). Consequently this reference to a connectivity is important and should be reflected in the IRs.

> Could you have a field with local fields that is not sparse? Do we need two definitions?

In theory sparse fields can be dense, but they never will be in practice. Nonetheless the concept remains valid so no two definitions are required. At this point a field defined on two locations which is dense is not considered, as it would violate the locality assumption behind what a stencil is, e.g. you can only access things in the vicinity of the currently considered location. For a spectral method we would however need this concept (and many more).

Reply:
I think I understand better now, but I think that, first we will not work on non-stencils, and second the term "sparse" has still no meaning to me. The "field defined on the neighbors of each grid point" sounds like a tensor of some kind? The think with the ordering is interesting.

Reply:
There is another important difference between a sparse field and a "normal" field with an extra dimension: the sparse dimension doesn't have fixed size. In the language of local field, on every grid point the local field could have different size.